### PR TITLE
Deprecate dispatch_main_async_safe, UIView+WebCache API now supports SDCallbackQueue and may callback in non-main queue

### LINKED ISSUE #3765: dispatch_main_async_safe issue
image

dispatch_main_async_safe 里面的判断不准，在使用 DiffableDataSource 的时候，Label 取出来跟 main thread 不一致，导致多 async 了一次。

### PR DESCRIPTION
**Pull request by dreampiggy**
### New Pull Request Checklist

* [ ] I have read and understood the [CONTRIBUTING guide](https://github.com/rs/SDWebImage/blob/master/.github/CONTRIBUTING.md)
* [ ] I have read the [Documentation](http://cocoadocs.org/docsets/SDWebImage/)
* [ ] I have searched for a similar pull request in the [project](https://github.com/rs/SDWebImage/pulls) and found none

* [ ] I have updated this branch with the latest master to avoid conflicts (via merge from master or rebase)
* [ ] I have added the required tests to prove the fix/feature I am adding
* [ ] I have updated the documentation (if necessary)
* [ ] I have run the tests and they pass
* [ ] I have run the lint and it passes (`pod lib lint`)

This merge request fixes / refers to the following issues: ...

### Pull Request Description

This close #3765

### CODE REVIEW

**Comment by dreampiggy** [comment]
Some one point this that we should always use `MainThread` to check for dispatch. Never use `MainQueue`

So what's the real correct way ? Why the SDWebImage original author use `dispatch_queue_main` to check if it's suitable to call the completion block in sync?

@Kyle-Ye

https://twitter.com/KyleSwifter/status/1856704385285566896

**Comment by dreampiggy** [comment]
## TLDR

For normal image loading, we use `SDCallbackPolicySafeAsyncMainQueue`, which only sync when in the main queue. This is for the common case. And I don't want to break exists user's behavior

For `UICollectionViewDiffableDataSource`, use `SDCallbackPolicySafeAsyncMainThread`, which don't care about GCD queue but only check main thread.
